app/jobs/plexer_job.rb

```diff
@@ -1,62 +1,54 @@
+# Preconditions:
+#  All needed PreservedCopy and ArchivePreservedCopy rows are already made.
+#  Possible TODO: replace precondition w/ invocation of PO2PC/PC2APC method.
+#
 # Responsibilities:
-# Record zip metadata info in DB.
+# Record zip part metadata info in DB.
 # Split message out to all necessary endpoints.
 # For example:
-#   Endpoint1Delivery.perform_later(druid, version)
-#   Endpoint2Delivery.perform_later(druid, version)
+#   Endpoint1Delivery.perform_later(druid, version, part_s3_key)
+#   Endpoint2Delivery.perform_later(druid, version, part_s3_key)
 #   ...
 #
 # Do not assume we can just get metadata from (the DruidVersionZip) zip.
-# Jobs are not run at the same time or on the same system, so the info will not match.
+# Jobs are not run at the same time or on the same system, so the info may not match.
 # Therefore, we receive the info passed by the process that was there when the file was created.
-class PlexerJob < DruidVersionJobBase
+class PlexerJob < ZipPartJobBase
   queue_as :zips_made
 
-  before_enqueue { |job| job.zip_info_check!(job.arguments.third) }
+  before_enqueue { |job| job.zip_info_check!(job.arguments.fourth) }
 
   # @param [String] druid
   # @param [Integer] version
+  # @param [String] part_s3_key, e.g. 'ab/123/cd/4567/ab123cd4567.v0001.z03'
   # @param [Hash<Symbol => String>] metadata Zip info
+  # @option metadata [Integer] :parts_count
   # @option metadata [Integer] :size
   # @option metadata [String] :checksum_md5
+  # @option metadata [String] :suffix
   # @option metadata [String] :zip_cmd
   # @option metadata [String] :zip_version
-  def perform(druid, version, metadata)
-    ApplicationRecord.transaction do
-      update_size(metadata[:size])
-      pcs.each do |pc|
-        pc.zip_checksums.find_or_create_by!(
-          md5: metadata[:checksum_md5],
-          create_info: metadata.slice(:zip_cmd, :zip_version).to_s
-        )
-      end
+  def perform(druid, version, part_s3_key, metadata)
+    apcs.each do |apc|
```

Comment:
I am confused here and I believe I understand the database schema.  

When I see "apcs" I'm not sure if there are multiple because
- there are multiple endpoints?  
- there are multiple parts for each ArchivedPreservedCopy?
- there are multiple version (I hope not, given that this job is for a single version)

The fact that I am confused, and that talking to John, it took him a few minutes to figure it out, to me implies that we need better names here.

Would it make  sense to move the first part into a method named something like "find_or_create_parts_for_apc" ?

Are we looping through apcs for endpoints, and then doing a separate set of gymnastics to figure out endpoints and then send a replication request for the single part to the endpoint?   Would it make more sense to do them together?  Eg.

```
apcs_by_endpoint do |apc|
   find_or_create_database_part
   endpoint_worker.perform_later
end
```

there is also confusion here because the database work seems to be by the individual parts for the zip, but the worker seems to not care about the individual parts.

Reply:
> there is also confusion here because the database work seems to be by the individual parts for the zip, but the worker seems to not care about the individual parts

i think plexer is only operating on one part of a zip at a time.  and so, `apcs` would be the `ArchivePreservedCopy` list for the druid/version combo (across all endpoints that are replication targets).  and the worker does care about individual parts, in that it only operates on the one part that was passed to the plexer worker (`worker.perform_later(druid, version, part_s3_key, metadata)` is operating on one part for one druid version for one endpoint).

i think if the `find_or_create_by` invocation were boxed up into its own method, the name would be `find_or_create_part_for_apc` (singular, because it will create one part, if necessary, for the given zip creation metadata and zip part).

i think that all of these things in conjunction lead to some parsing difficulty for humans:
* the loop iterates over the tersely named `apcs`.
* the `find_or_create_by` creates 0 or 1 records, but gets called on the pluralized field name `archive_preserved_copy_parts`.
* the fact that the `find_or_create_by` takes a multi-line hash for params, and then also takes a block to set one more thing on the newly created record (which took me a minute to parse even though i've written similar code before in this app).

i'd be fine with naomi's suggestion of maybe breaking out some of the sub-parts of perform into separate methods that feel more self-documenting.  i'd also be totally fine with a comment above the `apcs.each` along the lines of
```ruby
# iterates over archive preserved copies for the druid/version, one per archive endpoint.
# creates the `ArchivePreservedCopyPart` if needed for the given `part_s3_key` and `metadata`.
# queues a delivery job for the part for each target archive endpoint.
```

a bit more commentary at the class level might help too, indicating that plexer operates on a single part for a druid/version.

Reply:
the decision to have plexer operate on the level of a part instead of on the whole zip was a decision that was arrived at later in the week, and one that i forgot a couple times while reviewing this PR (including when naomi called me over to look at this part of the code, which was not helpful in reducing confusion for either of us).

not that that decision's being questioned here, but just to capture the reasoning for posterity:  from what @atz said, it was unwieldy to have zipmaker send over all the part info in one queue event, and having zipmaker create the DB records would introduce a database dependency we'd wanted to avoid for that worker.  we also realized that for 10 GB zip chunks, it might actually be nice to get the retries of the queuing mechanism for free on those sub-parts (as opposed to e.g. having a 50 GB cumulative upload get retried because one 10 GB part failed to upload).

this is also why results recorder operates at the level of a single part for a single endpoint now:  plexer plexes a single part to all endpoints, endpoint workers deliver one part to one endpoint at a time, and so a successful endpoint delivery worker can only queue a message for results recorder about the one part it just handled for its endpoint.